lgc/util/llpcGfxRegHelper.cpp

```diff
@@ -0,0 +1,311 @@
+// SqImgSampReg Bits infomation look up table (Gfx10)
+// Refer to TODO: update comment when the registers file is available
+static constexpr BitsInfo g_sqImgRsrcRegBitsGfx10[static_cast<uint32_t>(SqRsrcRegs::Count)] =
+{
+    { 0,  0, 32 }, // BaseAddress
+    { 1,  0,  8 }, // BaseAddressHi
+    { 1, 20,  9 }, // Format
+    {},            // Width
+    { 2, 14, 16 }, // Height
+    { 3,  0, 12 }, // DstSelXYZW
+    { 3, 20,  5 }, // IsTileOpt
+    { 4,  0, 16 }, // Depth
+    {},            // Pitch
+    { 3, 25,  3 }, // BcSwizzle
+    { 1, 30,  2 }, // WidthLo
+    { 2,  0, 14 }, // WidthHi
+};
+
+// =====================================================================================================================
+// Helper class for handling Registers defined in SQ_IMG_RSRC_WORD
+SqImgRsrcRegHelper::SqImgRsrcRegHelper(
+    Builder*      pBuilder,      // [in] Bound builder context
+    Value*        pRegister,     // [in] Bound register vec <n x i32>
+    GfxIpVersion* pGfxIpVersion) // [in] Current gfxIp version
+: GfxRegHelper(pBuilder, pRegister)
+{
+    m_pGfxIpVersion = pGfxIpVersion;
+
+    switch (pGfxIpVersion->major)
+    {
+    case 9:
+        m_bitsInfo = g_sqImgRsrcRegBitsGfx9;
+        break;
+    case 10:
+        m_bitsInfo = g_sqImgRsrcRegBitsGfx10;
+        break;
+    default:
+        llvm_unreachable("GfxIp is not supported!");
+        break;
+    }
+    SetBitsState(m_bitsState);
+}
+
+// =====================================================================================================================
+// Get the current llvm value for the hardware register
+Value* SqImgRsrcRegHelper::GetReg(
+    SqRsrcRegs id) // Register ID
+{
+    switch (id)
+    {
+    case SqRsrcRegs::BaseAddress:
+    case SqRsrcRegs::Format:
+    case SqRsrcRegs::DstSelXYZW:
+    case SqRsrcRegs::Depth:
+    case SqRsrcRegs::BcSwizzle:
+        return GetRegCommon(static_cast<uint32_t>(id));
+    case SqRsrcRegs::Height:
+    case SqRsrcRegs::Pitch:
+        return m_pBuilder->CreateAdd(GetRegCommon(static_cast<uint32_t>(id)), m_pInt32One);
+    case SqRsrcRegs::Width:
+        switch (m_pGfxIpVersion->major)
+        {
+        case 9:
+            return m_pBuilder->CreateAdd(GetRegCommon(static_cast<uint32_t>(id)), m_pInt32One);
+        case 10:
+            return m_pBuilder->CreateAdd(GetRegCombine(static_cast<uint32_t>(SqRsrcRegs::WidthLo),
+                                                       static_cast<uint32_t>(SqRsrcRegs::WidthHi)),
+                                                       m_pInt32One);
+        default:
+            llvm_unreachable("The current gfx is not supported!");
+            break;
+        }
+    case SqRsrcRegs::IsTileOpt:
+        return m_pBuilder->CreateICmpNE(GetRegCommon(static_cast<uint32_t>(id)), m_pBuilder->getInt32(0));
+    default:
+        llvm_unreachable("Not implemented!");
```

Comment:
The same as above
// TODO: More will be implemented.
llvm_unreachable("Not implemented!");
or
llvm_unreachable("Invalid register ID!");